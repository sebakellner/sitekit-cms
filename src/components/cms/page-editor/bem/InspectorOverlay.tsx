import React from 'react'
import type { ReactNode } from 'react'
import { Box } from 'grommet'
import { OverlayBorder } from './OverlayBorder'
import { OverlayLabel } from './OverlayLabel'
import { BEM_CHILD_ZINDEX } from './constants'

interface InspectorOverlayProps {
  label: string
  showOverlay?: boolean
  isSelected?: boolean
  onSelect?: () => void
  children: ReactNode
  overlayLabelPosition?: 'above' | 'below'
}

export const InspectorOverlay: React.FC<InspectorOverlayProps> = ({
  label,
  showOverlay = false,
  isSelected = false,
  onSelect,
  children,
  overlayLabelPosition = 'above',
}) => {
  return (
    <Box style={{ position: 'relative' }}>
      {showOverlay && (
        <OverlayBorder isSelected={isSelected} onSelect={onSelect} />
      )}
      {showOverlay && (
        <OverlayLabel
          label={label}
          isSelected={isSelected}
          position={overlayLabelPosition}
        />
      )}
      <Box style={{ position: 'relative', zIndex: BEM_CHILD_ZINDEX }}>
        {children}
      </Box>
    </Box>
  )
}
