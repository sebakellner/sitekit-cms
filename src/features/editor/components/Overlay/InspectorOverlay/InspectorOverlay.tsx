import React from 'react'
import { Box } from 'grommet'

import { BEM_CHILD_ZINDEX } from '@src/features/editor/constants/inspector'
import type { InspectorOverlayProps } from './InspectorOverlay.types'
import { OverlayBorder } from '../OverlayBorder'
import { OverlayLabel } from '../OverlayLabel'

const InspectorOverlay: React.FC<InspectorOverlayProps> = ({
  label,
  showOverlay = false,
  isSelected = false,
  onSelect,
  children,
  overlayLabelPosition = 'above',
}) => {
  return (
    <Box style={{ position: 'relative' }}>
      {showOverlay && (
        <OverlayBorder isSelected={isSelected} onSelect={onSelect} />
      )}
      {showOverlay && (
        <OverlayLabel
          label={label}
          isSelected={isSelected}
          position={overlayLabelPosition}
        />
      )}
      <Box style={{ position: 'relative', zIndex: BEM_CHILD_ZINDEX }}>
        {children}
      </Box>
    </Box>
  )
}

export default InspectorOverlay
