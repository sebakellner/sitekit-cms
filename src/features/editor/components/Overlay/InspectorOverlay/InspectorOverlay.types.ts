export interface InspectorOverlayProps {
  sectionId: string
  label: string
  showOverlay?: boolean
  isSelected?: boolean
  onSelect?: () => void
  children: React.ReactNode
  overlayLabelPosition?: 'above' | 'below'
}
