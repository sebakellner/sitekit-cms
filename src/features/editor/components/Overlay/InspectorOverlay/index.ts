export { default as InspectorOverlay } from './InspectorOverlay'
export type { InspectorOverlayProps } from './InspectorOverlay.types'
