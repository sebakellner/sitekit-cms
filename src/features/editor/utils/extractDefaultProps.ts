import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'
import type { ComponentMeta } from '@components/site/types'

function isDefaultProp(val: unknown): val is { default: unknown } {
  return typeof val === 'object' && val !== null && 'default' in val
}

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined,
  fallback: unknown = undefined
): Record<string, unknown> {
  if (!props) return {}
  const result = ComponentMetaSchema.shape.props.safeParse(props)
  if (!result.success) {
    console.error('Invalid props schema:', result.error)
    return {}
  }
  const validProps = result.data
  return Object.fromEntries(
    Object.entries(validProps).map(([key, val]) => [
      key,
      isDefaultProp(val) ? val.default : fallback,
    ])
  )
}
