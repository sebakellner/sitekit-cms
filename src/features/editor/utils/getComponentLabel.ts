import React from 'react'

export function getComponentLabel(
  children: React.ReactNode,
  fallback = 'Component'
): string {
  if (React.isValidElement(children)) {
    const type = children.type
    if (typeof type === 'function' && type.name) return type.name
  }
  return fallback
}
