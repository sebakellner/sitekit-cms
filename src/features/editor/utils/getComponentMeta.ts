import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'

type ModuleWithDefault = { default: unknown }

export function getComponentMeta(module: unknown) {
  if (
    module &&
    typeof module === 'object' &&
    'default' in module &&
    (module as ModuleWithDefault).default &&
    typeof (module as ModuleWithDefault).default === 'object'
  ) {
    const def = (module as ModuleWithDefault).default
    const result = ComponentMetaSchema.safeParse(def)
    if (result.success) {
      return result.data
    }
  }
  return { name: '', props: {} }
}
