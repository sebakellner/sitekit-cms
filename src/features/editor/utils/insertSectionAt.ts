import { componentsRegistry } from '@registry/componentRegistry'
import { v4 as uuidv4 } from 'uuid'

import type { ComponentMeta } from '@components/site/types'
import { extractDefaultProps } from '@src/features/editor/utils/extractDefaultProps'
import type { Section } from '../schemas/section.schema'

export async function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  name: string,
  index: number
) {
  const loader = componentsRegistry[name as keyof typeof componentsRegistry]
  if (!loader) {
    console.warn(`Component "${name}" not found in registry.`)
    return
  }

  const metaModule = await loader()

  const meta = metaModule.default as ComponentMeta

  const defaultProps = extractDefaultProps(meta?.props)

  const newSection: Section = {
    id: uuidv4(),
    name,
    props: defaultProps,
  }

  const updatedSections = [
    ...sections.slice(0, index),
    newSection,
    ...sections.slice(index),
  ]

  setSections(updatedSections)
}
