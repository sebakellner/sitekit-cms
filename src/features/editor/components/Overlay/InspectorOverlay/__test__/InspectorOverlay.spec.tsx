import { describe, test, expect } from 'vitest'
import { render } from 'test/utils/render'
import type { InspectorOverlayProps } from '../InspectorOverlay.types'
import InspectorOverlay from '../InspectorOverlay'

const setup = (overrides: Partial<InspectorOverlayProps> = {}) => {
  const defaultProps: InspectorOverlayProps = {
    sectionId: 'test-id',
    label: 'Component Test',
    showOverlay: true,
    isSelected: false,
    onSelect: vi.fn(),
    children: <div>Test Content</div>,
    overlayLabelPosition: 'above',
  }
  const props = { ...defaultProps, ...overrides }

  return <InspectorOverlay {...props} />
}

describe('InspectorOverlay', () => {
  test('should match with snapshot', () => {
    const { container } = render(setup())
    expect(container).toMatchSnapshot()
  })
})
