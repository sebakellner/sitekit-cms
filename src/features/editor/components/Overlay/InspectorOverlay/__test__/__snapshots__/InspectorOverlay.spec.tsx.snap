// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InspectorOverlay > should match with snapshot 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-79elbk"
  >
    <div
      class="css-kwi4bo"
      data-testid="overlay-border"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; z-index: 2; cursor: default; pointer-events: auto;"
    />
    <div
      class="css-vycisn"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-layers"
        fill="none"
        height="12"
        stroke="#007df0"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.83 2.18a2 2 0 0 0-1.66 0L2.6 6.08a1 1 0 0 0 0 1.83l8.58 3.91a2 2 0 0 0 1.66 0l8.58-3.9a1 1 0 0 0 0-1.83z"
        />
        <path
          d="M2 12a1 1 0 0 0 .58.91l8.6 3.91a2 2 0 0 0 1.65 0l8.58-3.9A1 1 0 0 0 22 12"
        />
        <path
          d="M2 17a1 1 0 0 0 .58.91l8.6 3.91a2 2 0 0 0 1.65 0l8.58-3.9A1 1 0 0 0 22 17"
        />
      </svg>
      <p
        class="css-1ti6hau"
      >
        Component Test
      </p>
    </div>
    <div
      class="css-rqgsqp"
    >
      <div>
        Test Content
      </div>
    </div>
  </div>
</div>
`;
